src/openrct2/core/String.cpp

```diff
@@ -343,66 +324,35 @@ namespace String
         return buffer;
     }
 
-    utf8* Format(const utf8* format, ...)
+    std::string StdFormat(const utf8* format, ...)
     {
         va_list args;
         va_start(args, format);
-        utf8* result = Format_VA(format, args);
+        std::string result = Format_VA(format, args);
         va_end(args);
         return result;
     }
 
-    utf8* Format_VA(const utf8* format, va_list args)
+    u8string Format_VA(const utf8* format, va_list args)
     {
-        va_list args1, args2;
-        va_copy(args1, args);
-        va_copy(args2, args);
-
-        // Try to format to a initial buffer, enlarge if not big enough
-        size_t bufferSize = 4096;
-        utf8* buffer = Memory::Allocate<utf8>(bufferSize);
+        va_list copy;
 
-        // Start with initial buffer
-        int32_t len = vsnprintf(buffer, bufferSize, format, args);
-        if (len < 0)
-        {
-            Memory::Free(buffer);
-            va_end(args1);
-            va_end(args2);
+        va_copy(copy, args);
 
-            // An error occurred...
-            return nullptr;
-        }
+        // Try to format to a initial buffer, enlarge if not big enough
+        int32_t len = vsnprintf(nullptr, 0, format, copy);
 
-        size_t requiredSize = static_cast<size_t>(len) + 1;
-        if (requiredSize > bufferSize)
-        {
-            // Try again with bigger buffer
-            buffer = Memory::Reallocate<utf8>(buffer, bufferSize);
-            len = vsnprintf(buffer, bufferSize, format, args);
-            if (len < 0)
-            {
-                Memory::Free(buffer);
-                va_end(args1);
-                va_end(args2);
+        va_end(copy);
 
-                // An error occurred...
-                return nullptr;
-            }
-        }
-        else
+        if (len >= 0)
         {
-            // Reduce buffer size to only what was required
-            bufferSize = requiredSize;
-            buffer = Memory::Reallocate<utf8>(buffer, bufferSize);
+            std::string buffer(std::size_t(len) + 1, '\0');
+            vsnprintf(buffer.data(), buffer.size(), format, args);
+            buffer.resize(len);
```

Comment:
No need to take the null terminator into account, std::string does not require it for its buffer. You can simplify this code to:

```suggestion
            std::string buffer(std::size_t(len), '\0');
            vsnprintf(buffer.data(), buffer.size() + 1, format, args);
```

Also, make len `const std::size_t`, so you do not have to cast it here.

Reply:
Also, std::size_t is unsigned, meaning that any encoding error never gets caught because a negative number never happens.

Reply:
Just confirmed earlier today with a small test project that vsnprintf does indeed write a null terminator at the end of a string, so std::string will hold two null terminators when c_str is called. Not really a problem, but it doesn't seem too nice to use a string as a buffer.